docs/developer-guide/join-streams-and-tables.rst

```diff
@@ -0,0 +1,329 @@
+late-arriving records. If a record arrives after the window’s retention period
+has passed, the record is discarded and isn’t processed in that window.
+
+Windows are tracked per record key. In join operations, KSQL uses a windowing
+*state store* to store all of the records received so far within the defined
+window boundary. Old records in the state store are purged after the specified
+window retention period. For more information, see Windows and Time [TBD in Concepts node].
+
+Join Requirements
+*****************
+
+Your KSQL applications must meet specific requirements for joins to be successful. 
+
+**Co-partitioned data**: Input data must be co-partitioned when joining. This ensures
+that input records with the same key, from both sides of the join, are delivered
+to the same stream task during processing. It’s your responsibility to ensure data
+co-partitioning when joining. For more info, see Partition Data to Enable Joins [TBD new topic].
+
+**KEY property**: If you set the KEY property in a KSQL query, ensure that both of the following
+conditions are true:
+
+* For every record, the contents of the message key of the Kafka message itself must be
+  the same as the contents of the column set in KEY.
+* The KEY property must be set to a column of type VARCHAR or STRING.
+
+TBD: code snippet
+
+For more information, see :ref:`ksql_key_requirements`.
+
+Join Capabilities
+*****************
+
+KSQL supports a large set of join operations for streams and tables, including
+INNER, LEFT OUTER, and FULL OUTER. Frequently, LEFT OUTER is shortened to LEFT JOIN,
+and FULL OUTER is shortened to OUTER JOIN.
+
+.. note:: RIGHT OUTER JOIN isn’t supported. Instead, swap the operands and use LEFT JOIN.
+
+The following table shows the supported combinations.
+
++---------------+--------------+-----------+------------+---------------+
+|               | Type         | INNER     | LEFT OUTER | FULL OUTER    |
++===============+==============+===========+============+===============+
+| Stream-Stream | Windowed     | Supported | Supported  | Supported     |                                         
++---------------+--------------+-----------+------------+---------------+
+| Table-Table   | Non-windowed | Supported | Supported  | Supported     |
++---------------+--------------+-----------+------------+---------------+
+| Stream-Table  | Non-windowed | Supported | Supported  | Not supported |
++---------------+--------------+-----------+------------+---------------+
+
+Stream-Stream Joins
+*******************
+
+KSQL supports INNER, LEFT OUTER, and FULL OUTER, joins between streams.
+
+All of these operations support out-of-order records.
+
+Joins between streams are always windowed joins. A new input record on one side
+produces a join output for each matching record on the other side, and there
+can be multiple such matching records within a join window.
+
+INNER joins cause data re-partitioning of a stream only if the stream was

```

Comment:
This repartitioning statement applies to all the join types, not just inner joins.